host/class/uac/usb_host_uac/uac_descriptors.c

```diff
@@ -0,0 +1,313 @@
+        uint32_t min_samfreq = (p_samfreq[2] << 16) + (p_samfreq[1] << 8) + p_samfreq[0];
+        uint32_t max_samfreq = (p_samfreq[5] << 16) + (p_samfreq[4] << 8) + p_samfreq[3];
+        printf("\ttLowerSamFreq %"PRIu32"\n", min_samfreq);
+        printf("\ttUpperSamFreq %"PRIu32"\n", max_samfreq);
+    } else {
+        const uint8_t *p_samfreq = desc->tSamFreq;
+        for (int i = 0; i < desc->bSamFreqType; ++i) {
+            printf("\ttSamFreq[%d] %"PRIu32"\n", i, (uint32_t)((p_samfreq[3 * i + 2] << 16) + (p_samfreq[3 * i + 1] << 8) + p_samfreq[3 * i]));
+        }
+    }
+}
+
+static void print_unknown_desc(const uac_desc_header_t *desc)
+{
+    printf("\t*** Unknown descriptor ***\n");
+    printf("\tbLength %d\n", desc->bLength);
+    printf("\tbDescriptorType 0x%x\n", desc->bDescriptorType);
+    printf("\tbDescriptorSubtype 0x%x\n", desc->bDescriptorSubtype);
+}
+
+static void print_uac_class_descriptors(const usb_standard_desc_t *desc, uint8_t subclass, uint8_t protocol)
+{
+    const uint8_t *buff = (const uint8_t *)desc;
+    uac_desc_header_t *header = (uac_desc_header_t *)desc;
+
+    if (subclass == UAC_SUBCLASS_AUDIOCONTROL) {
+        switch (header->bDescriptorSubtype) {
+        case UAC_AC_HEADER:
+            print_ac_header_desc(buff);
+            break;
+        case UAC_AC_INPUT_TERMINAL:
+            print_ac_input_desc(buff);
+            break;
+        case UAC_AC_OUTPUT_TERMINAL:
+            print_ac_output_desc(buff);
+            break;
+        case UAC_AC_FEATURE_UNIT:
+            print_ac_feature_desc(buff);
+            break;
+        default:
+            goto unknown;
+            break;
+        }
+    } else if (subclass == UAC_SUBCLASS_AUDIOSTREAMING && desc->bDescriptorType == UAC_CS_INTERFACE) {
+        switch (header->bDescriptorSubtype) {
+        case UAC_AS_GENERAL:
+            parse_as_general_desc(buff);
+            break;
+        case UAC_AS_FORMAT_TYPE:
+            parse_as_type_desc(buff);
+            break;
+        default:
+            goto unknown;
+            break;
+        }
+    } else if (subclass == UAC_SUBCLASS_AUDIOSTREAMING && desc->bDescriptorType == UAC_CS_ENDPOINT) {
+        switch (header->bDescriptorSubtype) {
+        case UAC_EP_GENERAL:
+            parse_as_ep_general_desc(buff);
+            break;
+        default:
+            break;
+        }
+    } else {
+        printf("\tUnknown subclass 0x%x\n", subclass);
+        goto unknown;
+    }
+    return;
+unknown:
+    print_unknown_desc(header);
+}
+
+// Print the configuration descriptor and all its sub-descriptors with the given class-specific callback
+// The subclass and protocol are passed to the class_specific_cb to allow it to interpret the descriptors more accurately
+// This function better be added to usb_helpers.c
+static void usb_print_config_descriptor_with_context(const usb_config_desc_t *cfg_desc, print_class_descriptor_with_context_cb class_specific_cb)
```

Comment:
We better to add the helper function to IDF, the Class driver need to know the interface context to parse the CS interface/endpoint which belong to it.

Reply:
The cb type is different, here defined a new callback function type with  `uint8_t subclass` as context

```
typedef void (*print_class_descriptor_with_context_cb)(const usb_standard_desc_t *, uint8_t subclass, uint8_t protocol);
```